### ISSUE TITLE
Vectors are not identical in `form`

### ISSUE THREAD

**Issue by srosenbu** [BAMresearch/fenics-constitutive#59]
https://github.com/BAMresearch/fenics-constitutive/blob/9285ebbd4af676cf829db6ecdf2f1d6575947194/src/fenics_constitutive/solver.py#L237-L239

So, it just happened to me that this assertion failed, which means that it can happen that the solution vecor supplied in the `form(self, x)` method is not the vector that has been created by the function `u`....

**Comment by joergfunger** [comment]
What exactly does that mean, why is that important, i.e. why was the assert added?

**Comment by srosenbu** [comment]
Sorry for the short description. I just wanted to add this issue as a reminder for myself. 

So basically, the solver from dolfinx provides the current solution vector with which we can determine the current strains. I assumed that this is the same vector that is inserted in the `solver.solve(u)` call. Furthermore, the `IncrSmallStrainProblem` requires the solution field `u` as input. Therefore the check was supposed to make sure that the `u` in `solver.solve(u)` and the one provided in the `IncrSmallStrainProblem` are the same (by checking the memory address) because from this `u` we create the expression that evaluates the gradients. This worked really well and we did not need to copy anything from the vector that dolfinx provided us because it was the vector from `u`.

However, in one simulation, the assert failed which meant that probably a copy of the solution vector was made somewhere, therefore we cannot work without copying the values anymore. I added this copy in https://github.com/BAMresearch/fenics-constitutive/pull/60
